test/distributed/org/apache/cassandra/distributed/test/streaming/StreamCloseInMiddleTest.java

```diff
@@ -0,0 +1,204 @@
+    {
+        // preemptive log search to make sure classes load... seems there is a bug where a downed instance can't return access to logs
+        cluster.forEach(n -> n.logs().grep("testing"));
+
+        IInvokableInstance node1 = cluster.get(1);
+        IInvokableInstance node2 = cluster.get(2);
+
+        // repair will do streaming IFF there is a mismatch; so cause one
+        for (int i = 0; i < 10; i++)
+            node1.executeInternal(withKeyspace("INSERT INTO %s.tbl (pk) VALUES (?)"), i);
+
+        // we do not have metrics for streaming impl; this is only done in debug level logging... so enable debug logging if not on
+        cluster.forEach(n -> n.nodetoolResult("setlogginglevel", "streaming", "DEBUG").asserts().success());
+
+        // trigger streaming; expected to fail as streaming socket closed in the middle (currently this is an unrecoverable event)
+        long mark = node2.logs().mark();
+        node2.nodetoolResult("repair", "-full", KEYSPACE, "tbl").asserts().failure();
+
+        assertStreamingFailed(node2, expectedEntireSSTable, mark);
+    }
+
+    private static void assertStreamingFailed(IInvokableInstance node, boolean expectedEntireSSTable, long mark)
+    {
+        // make sure zero-copy streaming was done; only way to know is by checking debug logs on the receiving end
+        LogResult<List<String>> logs = node.logs().grep(mark, "Incoming stream entireSSTable=");
+        Assertions.assertThat(logs.getResult()).isNotEmpty();
+        Pattern pattern = Pattern.compile("entireSSTable=(.*?)\\s");
+        logs.getResult().forEach(line -> {
+            Matcher match = pattern.matcher(logs.getResult().get(0));
+            Assertions.assertThat(match.find()).isTrue();
+            boolean entireSSTable = Boolean.parseBoolean(match.group(1));
+            Assertions.assertThat(entireSSTable).isEqualTo(expectedEntireSSTable);
+        });
+    }
+
+    public static class BBHelper
+    {
+        @SuppressWarnings("unused")
+        public static int writeDirectlyToChannel(ByteBuffer buf, @SuperCall Callable<Integer> zuper) throws Exception
+        {
+            if (isCaller(BigTableZeroCopyWriter.class.getName(), "write"))
+                throw new java.nio.channels.ClosedChannelException();
+            // different context; pass through
+            return zuper.call();
+        }
+
+        @SuppressWarnings("unused")
+        public static boolean append(UnfilteredRowIterator partition, @SuperCall Callable<Boolean> zuper) throws Exception
+        {
+            if (isCaller(CassandraIncomingFile.class.getName(), "read")) // handles compressed and non-compressed
+                throw new java.nio.channels.ClosedChannelException();
+            // different context; pass through
+            return zuper.call();
+        }
+
+        private static boolean isCaller(String klass, String method)
+        {
+            //TODO is there a cleaner way to check this?
```

Comment:
I think there's `StackWalker` in Java 9, but that's no help to us :)

Reply:
maybe in 10 years ^_^